internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,269 @@
+	b.WriteString(fmt.Sprintf("%sr%s/%s", QueryCachePrefix, store, tuple.TupleKeyToString(tupleKey)))
+	return c.newCachedIterator(ctx, iter, b.String())
+}
+
+type iterFunc func(ctx context.Context) (storage.TupleIterator, error)
+
+func (c *CachedDatastore) newCachedIterator(ctx context.Context, dsIterFunc iterFunc, key string) (storage.TupleIterator, error) {
+	span := trace.SpanFromContext(ctx)
+	tuplesCacheTotalCounter.Inc()
+
+	cacheKey := cacheKeyFor(key)
+	if cacheKey == "" {
+		return dsIterFunc(ctx)
+	}
+
+	cachedResp := c.cache.Get(cacheKey)
+	isCached := cachedResp != nil && !cachedResp.Expired && cachedResp.Value != nil
+
+	if isCached {
+		tuplesCacheHitCounter.Inc()
+		span.SetAttributes(attribute.Bool("cached", true))
+		return storage.NewStaticTupleIterator(cachedResp.Value.([]*openfgav1.Tuple)), nil
+	}
+
+	iter, err := dsIterFunc(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cachedIterator{
+		iter:          iter,
+		tuples:        make([]*openfgav1.Tuple, 0, c.maxResultSize),
+		cacheKey:      cacheKey,
+		cache:         c.cache,
+		maxResultSize: c.maxResultSize,
+		ttl:           c.ttl,
+	}, nil
+}
+
+// Close closes the datastore and cleans up any residual resources.
+func (c *CachedDatastore) Close() {
+	c.cache.Stop()
+	c.OpenFGADatastore.Close()
+}
+
+type cachedIterator struct {
+	iter          storage.TupleIterator
+	tuples        []*openfgav1.Tuple
+	cacheKey      string
+	cache         storage.InMemoryCache[any]
+	ttl           time.Duration
+	maxResultSize int
+	closeOnce     sync.Once
+	isClosed      bool
+}
+
+func (c *cachedIterator) addToBuffer(t *openfgav1.Tuple) bool {
+	if c.tuples == nil {
+		return false
+	}
+	c.tuples = append(c.tuples, t)
+	if len(c.tuples) >= c.maxResultSize {
+		c.tuples = nil
+	}
+	return true
+}
+
+// Next see [Iterator.Next].
+func (c *cachedIterator) Next(ctx context.Context) (*openfgav1.Tuple, error) {
+	t, err := c.iter.Next(ctx)
+	if err != nil {
+		if !errors.Is(err, storage.ErrIteratorDone) {
+			c.tuples = nil // don't store results that are incomplete
+		}
+		return nil, err
+	}
+
+	c.addToBuffer(t)
+
+	return t, nil
+}
+
+// Stop see [Iterator.Stop].
+func (c *cachedIterator) Stop() {
+	c.closeOnce.Do(func() {
+		if c.tuples == nil {
+			c.iter.Stop()
+			c.isClosed = true
+			return
+		}
+		// prevent goroutine if iterator was already consumed
+		ctx := context.Background()
+		if _, err := c.iter.Head(ctx); errors.Is(err, storage.ErrIteratorDone) {
+			c.flush()
+			c.iter.Stop()
+			c.isClosed = true
+			return
+		}
+
+		go func() {
+			defer c.iter.Stop()
+			defer func() {
+				c.isClosed = true
+			}()
+			for {
+				// attempt to drain the iterator to have it ready for subsequent calls
+				t, err := c.iter.Next(ctx)
+				if err != nil {
+					if errors.Is(err, storage.ErrIteratorDone) {
+						c.flush()
+					}
+					return
+				}
+				// if the size is exceeded we don't add anymore and exit
+				if !c.addToBuffer(t) {
+					return
+				}
+			}
+		}()
```

Comment:
If we are effectively closing the `cachedIterator` here, why do we need to drain the internal iterator? could we instead just dispose of any allocations in an efficient manner that could be done synchronously?

Reply:
oh, is the cache reused across instances?

Reply:
one thing to note is that after `Stop` completes, the caller will likely assume that the cache is in a stable state, however it is not, and may be modified at some undeterminable point in the future by a dangling go routine.